src/metrics/metric_manager.py

```diff
@@ -0,0 +1,177 @@
+
+        Return:
+            Metric list as nn.ModuleList for current phase. 
+        """
+        # create added_metric_names set
+        added_metric_names = set()
+        metrics = []
+        for metric_params in params:
+            if phase not in metric_params.phases:
+                continue
+            metric = METRICS.get(metric_params.class_name)(**metric_params.metric_params)
+            target_fields = metric_params.target_fields
+            name = metric_params.name if metric_params.name is not None else metric_params.class_name
+            if name in added_metric_names:
+                target_fields_values = list(target_fields.values())
+                name += + '_' + target_fields_values[0] + '_' + target_fields_values[1]
+            else:
+                added_metric_names.add(name)
+
+            metrics.append(MetricWithUtils(metric=metric, target_fields=target_fields, name=name))
+
+        metrics = nn.ModuleList(metrics)
+        return metrics
+
+    def forward(self, phase: str, *args, **kwargs):
+        """Update states of all metrics on phase loop.
+
+        Args:
+            phase: Phase name.
+        """
+        args = list(args)
+        if phase not in self.phases:
+            raise ValueError(f'Incorrect epoch setting. '
+                             f'Please choose one of {self.phases}')
+
+        for metric_with_utils in self.phase2metrics[phase]:
+            targeted_kwargs = self.map_arguments(metric_with_utils.target_fields, kwargs)
+            if targeted_kwargs:
+                # may be we need only update because forward use compute and sync all the processses
+                metric_with_utils(*args, **targeted_kwargs)
+            
+
+    def on_epoch_end(self, phase: str) -> Dict[str, Tensor]:
+        """Summarize epoch values and return log.
+        
+        Args:
+            phase: Run metric phase.
+
+        Returns:
+            Return logging dictionary, there the key is phase/metric_name and value is metric value on phase.
+
+        Raises:
+            ValueError: An error occure, when phase not in self.phases.
+            ValueError: An error occure, when metric.compute() return tensor with non zero shape.
+        """
+        if phase not in self.phases:
+            raise ValueError(f'Incorrect epoch setting. '
+                             f'Please choose one of {self.phases}')
+            
+        log = {}
+        for metric_with_utils in self.phase2metrics[phase]:
+            metric_value = metric_with_utils.compute()
+            if isinstance(metric_value, dict):
+                metric_value = list(metric_value.values())[0]
+            
+            if len(metric_value.shape) != 0:
+                raise ValueError(f'{metric_with_utils.name} must compute float value, '
+                                f'not torch tensor with shap {metric_value.shape}.')
+            
+            log[f'{phase}/{metric_with_utils.name}'] = metric_value
```

Comment:
Too long and complex key f'{phase}/{metric_with_utils.name}'.
Maybe:
metric_key = f'{phase}/{metric_with_utils.name}'
log[metric_key] = metric_value